datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/mirrormaker/KafkaMirrorMakerConnector.java

```diff
@@ -109,4 +157,193 @@ public void postDatastreamInitialize(Datastream datastream, List<Datastream> all
       throws DatastreamValidationException {
     _groupIdConstructor.populateDatastreamGroupIdInMetadata(datastream, allDatastreams, Optional.of(LOG));
   }
+
+  @Override
+  public void onPartitionChange(java.util.function.Consumer<DatastreamGroup> callback) {
+    if (!_enablePartitionAssignment) {
+      return;
+    }
+    _partitionChangeCallback = callback;
+  }
+
+  @Override
+  public void stop() {
+    super.stop();
+    _shutdown = true;
+    _partitionDiscoveryThreadMap.values().forEach(PartitionDiscoveryThread::shutdown);
+  }
+
+  /**
+   * Get the partitions for all datastream group. Return Optional.empty() for that datastreamGroup if the
+   * datastreamGroup has been assigned but the partition info has not been fetched already. This is only triggered
+   * in the LEADER_PARTITION_ASSIGNMENT thread so that it doesn't need to be thread safe.
+   */
+  @Override
+  public Map<String, Optional<DatastreamGroupPartitionsMetadata>> getDatastreamPartitions() {
+    Map<String, Optional<DatastreamGroupPartitionsMetadata>> datastreams = new HashMap<>();
+    _partitionDiscoveryThreadMap.forEach((s, partitionDiscoveryThread) -> {
+      if (partitionDiscoveryThread.isInitialized()) {
+        datastreams.put(s, Optional.of(new DatastreamGroupPartitionsMetadata(partitionDiscoveryThread.getDatastreamGroup(),
+            partitionDiscoveryThread.getSubscribedPartitions())));
+      } else {
+        datastreams.put(s, Optional.empty());
+      }
+    });
+    return datastreams;
+  }
+
+
+  /**
+   * callback when the datastreamGroups belongs this connector instance has been changed.
+   * This happens 1) when a new datastream is created/unpaused for this connector.
+   * 2) when a followers becomes a leader
+   *
+   * This is only triggered in the LEADER_DO_ASSIGNMENT thread so that it doesn't need to be thread safe.
+   */
+  @Override
+  public void handleDatastream(List<DatastreamGroup> datastreamGroups) {
+    if (!_enablePartitionAssignment || _partitionChangeCallback == null) {
```

Comment:
I think if the parttionChangeCallback is null, we should throw.

Reply:
make sense. I will throw an exception